src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs

```diff
@@ -46,23 +79,69 @@ private void Workspace_DocumentClosed(object sender, DocumentEventArgs e)
                     _openDocumentContexts.Remove(e.Document.Id);
 
                     // Ensure we dispose the context, which we'll do asynchronously
-                    contextTask.ContinueWith(
-                        t => t.Result.Dispose(),
-                        CancellationToken.None,
-                        TaskContinuationOptions.OnlyOnRanToCompletion,
-                        TaskScheduler.Default);
+                    EnsureContextCleanup(contextTask);
                 }
             }
         }
 
-        private void Workspace_DocumentOpened(object sender, DocumentEventArgs e)
+        private void OnDocumentOpened(object sender, DocumentEventArgs e)
         {
             lock (_gate)
             {
-                _openDocumentContexts.Add(e.Document.Id, Task.Run(() => GetConventionContextAsync(e.Document.FilePath, CancellationToken.None)));
+                var contextTask = Task.Run(async () =>
+                {
+                    var context = await GetConventionContextAsync(e.Document.FilePath, CancellationToken.None).ConfigureAwait(false);
+                    context.CodingConventionsChangedAsync += OnCodingConventionsChangedAsync;
+                    return context;
+                });
+
+                Contract.Requires(!_openDocumentContexts.ContainsKey(e.Document.Id));
+                _openDocumentContexts.Add(e.Document.Id, contextTask);
             }
         }
 
+        private void ClearOpenFileCache(ProjectId projectId = null)
+        {
+            var contextTasks = new List<Task<ICodingConventionContext>>();
+
+            lock (_gate)
+            {
+                if (projectId == null)
+                {
+                    contextTasks.AddRange(_openDocumentContexts.Values);
+                    _openDocumentContexts.Clear();
+                }
+                else
+                {
+                    foreach (var kv in _openDocumentContexts.Where(kv => kv.Key.ProjectId == projectId).ToList())
```

Comment:
would it make more sense to just key things off of project id, then doc id?

Reply:
@jasonmalinowski ?